src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -40,44 +39,49 @@ internal partial class TextDocumentState
 
         protected TextDocumentState(
             SolutionServices solutionServices,
+            IDocumentServiceProvider documentServiceProvider,
             DocumentInfo.DocumentAttributes attributes,
             SourceText sourceTextOpt,
             ValueSource<TextAndVersion> textAndVersionSource,
             ValueSource<DocumentStateChecksums> lazyChecksums)
         {
             this.solutionServices = solutionServices;
-            this.attributes = attributes;
             this.sourceTextOpt = sourceTextOpt;
             this.textAndVersionSource = textAndVersionSource;
 
+            Attributes = attributes;
+            Services = documentServiceProvider ?? DefaultDocumentServiceProvider.Instance;
+
             // for now, let it re-calculate if anything changed.
             // TODO: optimize this so that we only re-calcuate checksums that are actually changed
             _lazyChecksums = new AsyncLazy<DocumentStateChecksums>(ComputeChecksumsAsync, cacheResult: true);
         }
 
+        public DocumentInfo.DocumentAttributes Attributes { get; }
```

Comment:
I just changed the property to auto prop. not adding new one.

public DocumentInfo.DocumentAttributes Attributes	
        {	
            get { return this.attributes; }	
        }

Reply:
since state is internal, i could remove all those individual properites and just expose attribute. everytime we add something to attribute, adding new property in state seems just label. not sure what that improve.

Reply:
> I just changed the property to auto prop. not adding new one.

Agreed. The change can be reviewed in isolation in 5cd1dc03312e76d1a91f61832a2b8198f1a8be28. No change needed since before/after behavior is the same.